### ISSUE TITLE
`freeze()` BatchNorm eval mode is reset by the next model.train()

### ISSUE THREAD

**Issue by pgniewko** [pgniewko/gt-pyg#112]
Applies to `norm="bn"` only.

## Problem

`freeze()` sets `requires_grad=False` and puts frozen BatchNorm modules into eval mode, but that eval mode does not survive the next `model.train()`. Every training loop calls `model.train()` once per epoch, so frozen BatchNorm resumes training-mode behavior:

- `running_mean` / `running_var` keep updating and are overwritten by the fine-tuning data
- normalization uses current-batch statistics instead of the pretrained running statistics, so frozen-encoder outputs no longer match the checkpoint and each sample's embedding depends on the rest of its batch
- `get_frozen_status()` still returns `True`, and `save_checkpoint()` writes that `frozen_status` into the checkpoint, so the drift is not visible anywhere

## Cause

`gt_pyg/nn/model.py:356-367`:

```python
def _set_requires_grad(self, modules: List[nn.Module], requires_grad: bool) -> None:
    for module in modules:
        for param in module.parameters():
            param.requires_grad = requires_grad
        for m in module.modules():
            if isinstance(m, (nn.BatchNorm1d, nn.BatchNorm2d)):
                if requires_grad:
                    m.train()
                else:
                    m.eval()
```

Module mode is set once, at the moment `freeze()` / `unfreeze()` is called.
`nn.Module.train(mode)` recurses into every child and has no notion of frozen
state, so the next `model.train()` overwrites it.

BatchNorm buffer updates are driven by `module.training`, not by `requires_grad` or `torch.no_grad()`, so freezing the affine parameters does not stop the running statistics from moving.

Affected modules with `norm="bn"`: `input_norm` and `readout_norm` (`gt_pyg/nn/model.py`), `norm1`, `norm2`, `norm0e`, `norm1e` in every `GTConv` (`gt_pyg/nn/gt_conv.py:115-124`).

The same code has the inverse failure: `_set_requires_grad(..., True)` calls `m.train()` unconditionally, so `model.eval()` followed by `model.unfreeze(...)` leaves BatchNorm in training mode during validation.

## Reproduction

```python
m = GraphTransformerNet(..., norm="bn")
m.train()
m.freeze("encoder")
assert not m.input_norm.training      # passes
m.train()                             # what every epoch does
assert not m.input_norm.training      # fails
```

## Fix

Derive BatchNorm mode from frozen state on every mode switch instead of setting it at freeze time.

Remove the BatchNorm block from `_set_requires_grad`:

```python
def _set_requires_grad(self, modules: List[nn.Module], requires_grad: bool) -> None:
    """Set requires_grad for all params in modules."""
    for module in modules:
        for param in module.parameters():
            param.requires_grad = requires_grad
```

Add a `train()` override to `GraphTransformerNet`:

```python
def train(self, mode: bool = True) -> "GraphTransformerNet":
    """Set training mode, keeping frozen BatchNorm layers in eval."""
    super().train(mode)
    if mode:
        for m in self.modules():
            if isinstance(m, nn.modules.batchnorm._BatchNorm):
                params = list(m.parameters(recurse=False))
                if params and not any(p.requires_grad for p in params):
                    m.eval()
    return self
```

End `freeze()` and `unfreeze()` with `self.train(self.training)` so the change takes effect immediately and in both directions:

```python
    self._set_requires_grad(list(to_freeze), requires_grad=False)
    self.train(self.training)
    return self
```

`affine=False` BatchNorm has no parameters to read frozen state from and is left untouched; `make_norm` (`gt_pyg/nn/utils.py:28-31`) builds affine BatchNorm, so this case does not arise here.

## Tests

`test_batchnorm_eval` (`gt_pyg/nn/tests/test_model.py:131`) asserts immediately after `freeze()` and never calls `train()` again, so it passes against the current behavior. Required cases:

- `train()` → `freeze("encoder")` → `train()` → BatchNorm still in eval
- forward pass after that sequence leaves `running_mean` unchanged
- `eval()` → `unfreeze("heads")` → BatchNorm still in eval

## Downstream

`pgniewko/golem` uses `norm="bn"` by default (`golem/config.py:21`) and calls `model.train(is_training)` per epoch (`golem/pretrain.py:306`), but never calls `freeze()`, so its pretraining runs are unaffected. Fine-tuning a golem checkpoint with a frozen encoder hits this.
